roadmap/implementers-guide/src/runtime/session_info.md

```diff
@@ -0,0 +1,55 @@
+# Session Info
+
+For disputes and approvals, we need access to information about validator sets from prior sessions. We also often want easy access to the same information about the current session's validator set. This module aggregates and stores this information in a rolling window while providing easy APIs for access.
+
+## Storage
+
+Helper structs:
+
+```rust
+struct SessionInfo {
+    // validators in canonical ordering.
+    validators: Vec<ValidatorId>,
+    // validators' authority discovery keys for the session in canonical ordering.
+    discovery_keys: Vec<DiscoveryId>,
+    // The assignment and approval keys for validators.
+    approval_keys: Vec<(AssignmentId, ApprovalId)>,
+    // validators in shuffled ordering
+    validator_groups: Vec<Vec<ValidatorIndex>>,
```

Comment:
what's the meaning of the index of the outer variant? what does it correspond to?

Reply:
I'm not certain, but AFAIK the outer index doesn't mean anything in particular; it's just a way to separate the various groups of sets of validators.